src/helpers.rs

```diff
@@ -1015,6 +1015,67 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
         )
     }
 
+    // Creates Expected FnAbi of the called fn from the args and matches it's compatability with the
+    // FnAbi passed along miri the shim handling and also checks for mismatch between shim symbol and abi.
+    fn check_shim_abi<'a, const N: usize>(
+        &mut self,
+        link_name: Symbol,
+        caller_fn_abi: &FnAbi<'tcx, Ty<'tcx>>,
+        abi: ExternAbi,
+        callee_input_args_ty: &[Ty<'tcx>],
+        callee_output_ty: Ty<'tcx>,
+        caller_args: &'a [OpTy<'tcx>],
+    ) -> InterpResult<'tcx, &'a [OpTy<'tcx>; N]>
+    where
+        &'a [OpTy<'tcx>; N]: TryFrom<&'a [OpTy<'tcx>]>,
+    {
+        let this = self.eval_context_mut();
+        let mut inputs_and_output = callee_input_args_ty.to_vec();
+        inputs_and_output.push(callee_output_ty);
+        let fn_sig_binder = Binder::dummy(FnSig {
+            inputs_and_output: this.machine.tcx.mk_type_list(&inputs_and_output),
+            c_variadic: false,
+            safety: Safety::Safe,
+            abi,
+        });
+        let callee_fn_abi = this.fn_abi_of_fn_ptr(fn_sig_binder, Default::default())?;
+
+        let caller_args_abi = &caller_fn_abi.args;
+        let callee_args_abi = &callee_fn_abi.args;
+
+        if caller_fn_abi.c_variadic != callee_fn_abi.c_variadic
+            || callee_fn_abi.fixed_count != caller_fn_abi.fixed_count
+            || callee_fn_abi.can_unwind != caller_fn_abi.can_unwind
+            || callee_args_abi.len() != callee_args_abi.len()
+            || !this.check_argument_compat(&caller_fn_abi.ret, &callee_fn_abi.ret)?
+            || !caller_args_abi
+                .iter()
+                .zip(callee_args_abi.iter())
+                .map(|(caller_arg, callee_arg)| this.check_argument_compat(caller_arg, callee_arg))
+                .collect::<InterpResult<'tcx, Vec<bool>>>()?
+                .into_iter()
+                .all(|b| b)
+        {
+            throw_ub_format!("Invalid argument type: ABI mismatch");
+        }
```

Comment:
Instead of one big `if`, please do separate checks so that we can give an error that says a bit more precisely what is going wrong.